dev-tools/releng/src/hubris.rs

```diff
@@ -0,0 +1,145 @@
+        if let Some(hash) = line.split_whitespace().next() {
+            let data = fetch_hash(base_url, &client, hash).await?;
+            let str = String::from_utf8(data).with_context(|| {
+                format!("hubris artifact manifest {} was not UTF-8", hash)
+            })?;
+            let hash_manifest: Manifest =
+                toml::from_str(&str).with_context(|| {
+                    format!(
+                        "failed to deserialize hubris artifact manifest {}",
+                        hash
+                    )
+                })?;
+            for (kind, artifacts) in hash_manifest.artifacts {
+                for artifact in artifacts {
+                    let (source, hashes) = match artifact.source {
+                        Source::File(file) => (
+                            DeserializedArtifactSource::File {
+                                path: zip!(file.hash),
+                            },
+                            vec![file.hash],
+                        ),
+                        Source::CompositeRot { archive_a, archive_b } => (
+                            DeserializedArtifactSource::CompositeRot {
+                                archive_a:
+                                    DeserializedFileArtifactSource::File {
+                                        path: zip!(archive_a.hash),
+                                    },
+                                archive_b:
+                                    DeserializedFileArtifactSource::File {
+                                        path: zip!(archive_b.hash),
+                                    },
+                            },
+                            vec![archive_a.hash, archive_b.hash],
+                        ),
+                    };
+                    manifest.artifacts.entry(kind).or_default().push(
+                        DeserializedArtifactData {
+                            name: artifact.name,
+                            version: artifact.version,
+                            source,
+                        },
+                    );
+                    for hash in hashes {
+                        let data = fetch_hash(base_url, &client, &hash).await?;
+                        fs::write(output_dir.join(zip!(hash)), data).await?;
+                    }
+                }
+            }
+        }
+    }
+
+    fs::write(
+        output_dir.join("manifest.toml"),
+        toml::to_string_pretty(&manifest)?.into_bytes(),
+    )
+    .await?;
+    Ok(())
+}
+
+async fn fetch_hash(
+    base_url: &'static str,
+    client: &reqwest::Client,
+    hash: &str,
+) -> Result<Vec<u8>> {
+    client
+        .get(format!("{}/artifact/{}", base_url, hash))
+        .send()
+        .and_then(|response| response.json())
+        .await
+        .with_context(|| {
+            format!(
+                "failed to fetch hubris artifact {} from {}",
+                hash, base_url
+            )
+        })
```

Comment:
it's unfortunate that permslip is still a private repo but I guess it's mostly just a wrapper around reqwest anyway

Reply:
In an earlier revision (not pushed) I ended up copying the openapi.json out of the repo and generating a client with Progenitor, and then I realized how much that was slowing down rust-analyzer just for a single API call. So I went with this relatively-small wrapper.